tft.py

```diff
@@ -0,0 +1,189 @@
+        return other.__dict__
+
+
+class TFTTransformer:
+    def parse(self, input):
+        map22_file = os.path.join(input, "data", "maps", "shipping", "map22", "map22.bin")
+        character_folder = os.path.join(input, "data", "characters")
+
+        map22 = BinFile(map22_file)
+        items = self.parse_items(map22)
+        traits = self.parse_traits(map22)
+        champs = self.parse_champs(map22, traits, character_folder)
+
+        # clean up data for JSON export
+        [x.pop("internal") for x in traits]
+        [x.pop("internal") for x in items]
+
+        self.data = {"items": items, "traits": traits, "champs": champs}
+
+    def export(self, output, font_files=[], output_template=False):
+        os.makedirs(output, exist_ok=True)
+        template = json.dumps(self.data, cls=NaiveJsonEncoder, indent=4, sort_keys=True)
+
+        if output_template:
+            with open(os.path.join(output, "template.json"), "w", encoding="utf-8") as f:
+                f.write(template)
+
+        for language in font_files:
+            use = template
+            code = os.path.basename(language)[11:-4]
+
+            with open(language, "r", encoding="utf-8") as f:
+                for line in f:
+                    if line.startswith("tr"):
+                        key, val = line.split("=", 1)
+                        key = key[4:-2]
+                        val = val[2:-2]
+                        use = use.replace(key, val)
+
+            with open(os.path.join(output, code + ".json"), "w", encoding="utf-8") as f:
+                f.write(use)
+
+    def parse_items(self, map22):
+        item_collection = [x for x in map22.entries if x.type == "TftItemData"]
+        items = []
+
+        for item in item_collection:
+            name = item.getv("mName")
+            if "Template" in name or name == "TFT_Item_Null":
+                continue
+
+            effects = []
+            for effect in item.getv("effectAmounts") if item.get("effectAmounts") else []:
+                effects.append({"name": effect.getv("name"), "value": effect.getv("value")})
+
+            items.append(
+                {
+                    "internal": {"hex": item.path.hex(), "name": item.getv("mName")},
+                    "id": item.getv("mId"),
+                    "name": item.getrv("c3143d66"),
+                    "desc": item.getrv("765f18da"),
+                    "icon": item.getv("mIconPath"),
+                    "from": [x.hex() for x in item.getrv("8b83ba8a")] if item.getr("8b83ba8a") else [],
+                    "effects": effects,
+                }
+            )
+
+        # replace the hex key gotten earlier with the id
+        for item in items:
+            # there must be some list comprehension way of doing this...
+            for old in item["from"]:
+                key = [x for x in items if x["internal"]["hex"] == old][0]["id"]
+                item["from"].insert(0, key)
+                item["from"].remove(old)
```

Comment:
You cannot reliabily iterate on a list you are updating at the same time. It won't crash but you may obtain unwanted results. I guess you made it work with your `insert()`/`remove()` but I don't think you can safely rely on current behavior.

Also, regarding your comment, you could make `items` a dict for easier lookup. That would also solve the "update and iterate" problem.
```py
items = {}
for item in item_collection:
    ...
    items[item.path.h] = {
        "internal": {"hex": item.path.h, "name": ...},
        ...
    }

# convert "from" values from hex to id
for item in items.values():
    item["from"] = items[item["from"]]["id"]

return items.values()
```

Reply:
`item["from"] is a list, so id still need to .append tho?`

Reply:
Oh sorry, use: `item["from"] = [items[x]["id"] for x in items["from"]]`